### DISCUSSION TITLE
leaks reported from sphinx doctest

### DISCUSSION THREAD

**Discussion by wojdyr** [wjakob/nanobind#717] [Q&A] [answered]
When running `sphinx-build -M doctest . _build` on docs from my projects I often see a leak report like this:
```
nanobind: leaked 10 instances!
 - leaked instance 0x7fba40aff708 of type "WriteOptions"
 - leaked instance 0x7fba3fad2508 of type "Element"
…
nanobind: leaked 59 types!
 - leaked type "gemmi.ResidueGroup"
 - leaked type "gemmi.Residue"
…
nanobind: leaked 792 functions!
 - leaked function ""
 - leaked function ""
 - leaked function "__getitem__"
 - leaked function "__repr__"
…
```
It's not deterministic and the reports vary.
I get it from various sphinx versions, including the latest one, on all systems (Linux, Windows, Mac).
No leaks are reported when running other tests and programs.

Anyone else has this problem?
Any idea how to avoid it or investigate it?

**Comment by wjakob** [comment, marked as answer]
There is a [FAQ entry](https://nanobind.readthedocs.io/en/latest/faq.html#why-am-i-getting-errors-about-leaked-functions-and-types) about these kinds of leaks. At this point, I've several times gone down the rabbit hole of back-tracking the source, and it is usually some issue in another library or an issue with Python GC not being comprehensive enough at shutdown time. They are actual leaks, so it is correct that nanobind reports them, but they might not be serious enough to show to users of your library.

You can disable these warnings if they bother you. I keep them on by default since it's still useful to know if there are some serious problems.
